warehouse/manage/views/__init__.py

```diff
@@ -1435,6 +1436,117 @@ def manage_project_oidc_publishers_prefill(self):
+
+        if not form.validate():
+            self.request.session.flash(
+                self.request._("The trusted publisher could not be constrained"),
+                queue="error",
+            )
+            return self.default_response
+
+        publisher = self.request.db.get(OIDCPublisher, form.publisher_id.data)
+
+        if publisher is None or publisher not in self.project.oidc_publishers:
+            self.request.session.flash(
+                "Invalid publisher for project",
+                queue="error",
+            )
+            return self.default_response
+
+        # First we add the new trusted publisher
+        if isinstance(publisher, GitHubPublisher):
+            constrained_publisher = GitHubPublisher(
+                repository_name=publisher.repository_name,
+                repository_owner=publisher.repository_owner,
+                repository_owner_id=publisher.repository_owner_id,
+                workflow_filename=publisher.workflow_filename,
+                environment=form.constrain_environment.data,
+            )
+        elif isinstance(publisher, GitLabPublisher):
+            constrained_publisher = GitLabPublisher(
+                namespace=publisher.namespace,
+                project=publisher.project,
+                workflow_filepath=publisher.workflow_filepath,
+                environment=form.constrain_environment.data,
+            )
+
+        else:
+            self.request.session.flash(
+                "Can only constrain the environment for GitHub and GitLab publishers",
+                queue="error",
+            )
+            return self.default_response
+
+        if publisher.environment != "":
+            self.request.session.flash(
+                "Can only constrain the environment for publishers without an "
+                "environment configured",
+                queue="error",
+            )
+            return self.default_response
+
+        self.request.db.add(constrained_publisher)
+        self.request.db.flush()  # ensure constrained_publisher.id is available
+        self.project.oidc_publishers.append(constrained_publisher)
+
+        self.project.record_event(
+            tag=EventTag.Project.OIDCPublisherAdded,
+            request=self.request,
+            additional={
+                "publisher": constrained_publisher.publisher_name,
+                "id": str(constrained_publisher.id),
+                "specifier": str(constrained_publisher),
+                "url": constrained_publisher.publisher_url(),
+                "submitted_by": self.request.user.username,
+            },
+        )
+
+        # Then, we remove the old trusted publisher from the project
+        # and, if there are no projects left associated with the publisher,
+        # we delete it entirely.
+        self.project.oidc_publishers.remove(publisher)
+        if len(publisher.projects) == 0:
+            self.request.db.delete(publisher)
```

Comment:
I might be missing something, but can we simplify this by modifying the publisher rather than creating a new one and deleting the old one?

Reply:
It wouldn't cover the case where the publisher is associated to more than one project: since the magic link is associated to a single project, it should only constraint the publisher for that project. And modifying a publisher would modify it for all associated projects.

Reply:
Ah right, completely forgot about that case! It might be good to leave a comment to that effect (the current comment explains it indirectly, but an explicit one might stop an over-eager refactor 🙂)